### PR TITLE
Quorum rotation (phase 1)

### PR DESCRIPTION
**Pull request by HashEngineering**
## Issue being fixed or feature implemented

This provides the skeleton and basic tests for quorum rotation

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

File: core/src/main/java/org/bitcoinj/evolution/SimplifiedMasternodeListManager.java [hunk 1]
```diff
@@ -811,18 +1057,416 @@ public SimplifiedQuorumList getQuorumListForBlock(Sha256Hash blockHash) {
...
+            }
+            for (SimplifiedMasternodeListEntry m: previousQuarters.quarterHMinusC.get(i)) {
+                for (int j = 0; j < quorumMembers.get(i).size(); ++j) {
+                    if (m.equals(quorumMembers.get(i).get(j))) {
+                        log.info("{} is already in the list", m);
+                    }
+                }
+                quorumMembers.get(i).add(m);
+            }
+            for (SimplifiedMasternodeListEntry m: newQuarterMembers.get(i)) {
```
Comment (Syn-McJ) [inline core/src/main/java/org/bitcoinj/evolution/SimplifiedMasternodeListManager.java:1179]: Repeating code - might be better to extract into a function.

File: core/src/main/java/org/bitcoinj/evolution/SimplifiedMasternodeListManager.java [hunk 2]
```diff
@@ -954,4 +1598,16 @@ public ReentrantLock getLock() {
...
+            cache.put(llmq.getKey(), new HashMap<>(llmq.getValue().getSigningActiveQuorumCount() +1));
+        }
+    }
+
+    void initQuorumsCache(HashMap<LLMQParameters.LLMQType, HashMap<Sha256Hash, ArrayList<Masternode>>> cache) {
+        for (Map.Entry<LLMQParameters.LLMQType, LLMQParameters> llmq : params.getLlmqs().entrySet()) {
+            cache.put(llmq.getKey(), new HashMap<>(llmq.getValue().getSigningActiveQuorumCount() +1));
+        }
+    }
 }
```
Comment (Syn-McJ) [inline core/src/main/java/org/bitcoinj/evolution/SimplifiedMasternodeListManager.java:1613]: Quite a large class - might be beneficial to break it up somehow.

**Review by Syn-McJ — APPROVED** [review]
Looks great, I have left a couple of comments. There is also some commented code in `SimplifiedMasternodeListManager.java` - maybe better to leave a TODO if it's still needed.

File: core/src/main/java/org/bitcoinj/evolution/SimplifiedMasternodeListManager.java (hunk 2 above)
Comment (HashEngineering) [reply core/src/main/java/org/bitcoinj/evolution/SimplifiedMasternodeListManager.java:1613]: I will see how I can do this in the next phase.

File: core/src/main/java/org/bitcoinj/evolution/SimplifiedMasternodeListManager.java (hunk 1 above)
Comment (HashEngineering) [reply core/src/main/java/org/bitcoinj/evolution/SimplifiedMasternodeListManager.java:1179]: Good idea. part of this code is to check for duplicates in lists (which is a bug I am trying to find the cause of).

**Pull request merged**